File: app/static/facebook.py
from facebook import GraphAPI
import datetime as dt
from json import loads
from requests import get
from re import search

# Exchange short lived access token for long lived access token
def long_token(short_token):
    app_id = "667185007475640"
    app_secret = "a6e48db00cf8e1fb260bebbf3d56a6ab"
    query = "https://graph.facebook.com/oauth/access_token?grant_type=fb_exchange_token&client_id={}&client_secret={}&fb_exchange_token={}".format(app_id, app_secret, short_token)
    response = get(query)
    content = response.content
    token = json.loads(content.decode('utf-8'))['access_token']
    return token

def create_graph(token):
    return GraphAPI(access_token=token, version="3.1")

# Pages this user account has access to
def get_pages(graph):
    pages = {}
    pages_info = graph.get_object(id="me", fields="accounts")
    pages_info = pages_info['accounts']['data']

    for i, v in enumerate(pages_info):
        page_name = pages_info[i]['name']
        page_id = pages_info[i]['id']
        pages[page_name] = page_id

    return pages

# Get the UTMs from an URL and save it as an object
def get_utms(url):
    utm_source = search("utm_source%3D(.*?)%", url).group(1) if search("utm_source%3D(.*?)%", url) else None
    utm_medium = search("utm_medium%3D(.*?)%", url).group(1) if search("utm_medium%3D(.*?)%", url) else None
    utm_campaign = search("utm_campaign%3D(.*?)%", url).group(1) if search("utm_campaign%3D(.*?)%", url) else None
    utm_content = search("utm_content%3D(.*?)%", url).group(1) if search("utm_content%3D(.*?)%", url) else None
    utm_term = search("utm_term%3D(.*?)%", url).group(1) if search("utm_term%3D(.*?)%", url) else None

    utms = {
        "utm_source": utm_source,
        "utm_medium": utm_medium,
        "utm_campaign": utm_campaign,
        "utm_content": utm_content,
        "utm_term": utm_term
    }

    return utms

# List of posts from the page
def get_posts(graph, page, start_date=0, end_date=0): # Expect dates in YYYY-MM-DD format

    # Convert start_date and end_date if provided, else use today's date
    start_date = dt.date.today() if start_date == 0 else dt.datetime.strptime(start_date, "%Y-%m-%d").date()
    end_date = dt.date.today() if end_date == 0 else dt.datetime.strptime(end_date, "%Y-%m-%d").date()

    # Get first page of posts
    posts = graph.get_object(id=page + "/posts", fields="created_time,permalink_url,id,attachments")

    first_match = False # Found a first post within the date range.
    last_match = False # Found a first post outside the date range, after finding a first_match
    filtered = [] # List of posts within date range

    i = 0 # Temporary to not end up in an endless loop

    while last_match == False and i < 10:
        i += 1

        # Only append posts created within the daterange
        for post in posts['data']:
            created = dt.datetime.strptime(post['created_time'], "%Y-%m-%dT%H:%M:%S+0000").date()

            if created >= start_date and created <= end_date:
                post_transformed = {
                    "id": post["id"],
                    "created_time": post["created_time"],
                    "url": post["permalink_url"],
                    "target": post['attachments']['data'][0]['target']['url'],
                    "utms": get_utms(post['attachments']['data'][0]['target']['url'])
                }

                filtered.append(post_transformed)
                first_match = True

            elif first_match == True:
                last_match = True

        # Get the next page of results
        if last_match == False:
            next_page_query = posts['paging']['next']
            next_page_response = get(next_page_query)
            posts = loads(next_page_response.content.decode("utf-8"))

    return filtered

File: app/static/js/facebook.js
let authorised = undefined;
let pages = {};
//let posts = undefined;
let page = undefined;
let next = undefined;

window.fbAsyncInit = function() {
  FB.init({
    appId: '2769812323343725',
    cookie: true,
    xfbml: false,
    version: 'v8.0'
  });

  FB.getLoginStatus(function(response) {
    statusChangeCallback(response);
  });

};

(function(d, s, id) {
  var js, fjs = d.getElementsByTagName(s)[0];
  if (d.getElementById(id)) {
    return;
  }
  js = d.createElement(s);
  js.id = id;
  js.src = "https://connect.facebook.net/en_US/sdk.js";
  fjs.parentNode.insertBefore(js, fjs);
}(document, 'script', 'facebook-jssdk'));

function statusChangeCallback(response) {
  if (response.status == 'connected') {
    authorised = true
    // Check for the right permissions
    getPages()
  } else {
    authorised = false;
    // Hide everything else until someone has logged in and given the right permissions
  }
}

function checkLoginState() {
  FB.getLoginStatus(function(response) {
    statusChangeCallback(response);
  });
}

function getPages() {
  FB.api('/me?fields=accounts', function(response) {
    if (response && !response.error) {
      const response_pages = response['accounts']['data'];
      const select = document.getElementById("pages__list");

      for (page of response_pages) {
        if (!(page['name'] in pages)) {
          pages[page['name']] = {
            'id': page['id'],
            'access_token': page['access_token']
          };
          const opt = document.createElement('option');
          opt.innerText = page['name'];
          opt.pageId = page['id'];
          opt.accessToken = page['access_token'];
          select.appendChild(opt);
        }
      }
      return pages;
    }
  })
}

function getUtms(url) {
  const source = url.match(/utm_source%3D(.*?)%/)
  const medium = url.match(/utm_medium%3D(.*?)%/)
  const campaign = url.match(/utm_campaign%3D(.*?)%/)
  const content = url.match(/utm_content%3D(.*?)%/)
  const term = url.match(/utm_term%3D(.*?)%/)

  utms = {
    "source": source ? source[1] : "-",
    "medium": medium ? medium[1] : "-",
    "campaign": campaign ? campaign[1] : "-",
    "content": content ? content[1] : "-",
    "term": term ? term[1] : "-"
  }
  return utms
}

function formatPosts(posts) {
  for (const post of posts) {
    const hasAttachments = Object.keys(post).includes('attachments')

    formattedPost = {
      "created_time": post["created_time"].slice(0,10) + " " + post["created_time"].slice(11,16),
      "fbUrl": post["permalink_url"],
      "outboundUrl": hasAttachments ? (post['attachments']['data'][0]['target']['url'].includes("l.facebook") ? post['attachments']['data'][0]['target']['url'] : undefined) : undefined,
      "utms": hasAttachments ? getUtms(post['attachments']['data'][0]['target']['url']) : {"source": "-", "medium": "-", "campaign": "-", "content": "-", "term": "-"}
    }

    const table = document.getElementById("posts__table");

    if(!document.getElementById('posts__table__body')) {
      const tableBody = document.createElement("tbody");
      tableBody.id = "posts__table__body";
      table.append(tableBody);
    }

    const tableBody = document.getElementById("posts__table__body");
    let row = tableBody.insertRow(-1);
    let cellCreatedAt = row.insertCell(0);
    let cellFacebookUrl = row.insertCell(1);
    let cellOutboundUrl = row.insertCell(2);
    let cellUtmSource = row.insertCell(3);
    let cellUtmMedium = row.insertCell(4);
    let cellUtmCampaign = row.insertCell(5);
    let cellUtmTerm = row.insertCell(6);
    let cellUtmContent = row.insertCell(7);

    row.className = "posts__row";
    cellCreatedAt.innerHTML = formattedPost.created_time; // Formatteer naar datum en tijd, zonder timezone.
    cellFacebookUrl.innerHTML = `<a href="${formattedPost.fbUrl}">Link</a>`; // Gebruik het Facebook logo.
    cellOutboundUrl.innerHTML = formattedPost.outboundUrl ? `<a href="${formattedPost.outboundUrl}">Link</a>` : "-"; // Gebruik een href icon.
    cellUtmSource.innerHTML = formattedPost.utms.source;
    cellUtmMedium.innerHTML = formattedPost.utms.medium;
    cellUtmCampaign.innerHTML = formattedPost.utms.campaign;
    cellUtmContent.innerHTML = formattedPost.utms.content;
    cellUtmTerm.innerHTML = formattedPost.utms.term;
  }
}

function getPosts() {
  const select = document.getElementById("pages__list");
  const pageName = select.options[select.selectedIndex].innerText;

  if (document.getElementById("posts__table__body")) {
      const tableBody = document.getElementById("posts__table__body");
      tableBody.remove()
    }

  let page_id = pages[pageName]['id']
  FB.api(`/${page_id}/published_posts`, {
      'access_token': pages[pageName]['access_token'],
      'fields': 'created_time,permalink_url,id,attachments'
    },

    function(response) {
      posts = response['data']
      next = response['paging']['next'] ? response['paging']['next'] : undefined
      formatPosts(posts)
    })
}

let gnp = undefined
function getNext() {
  fetch(next)
  .then(function(response) {
    return response.json()
  })
  .then(function(data) {
    gnp = data
    let posts = data['data']
    formatPosts(posts)

    next = data['paging']['next'] ? data['paging']['next'] : undefined
  })
}
